### PR TITLE
feat(empty screen): new empty message component

### LINKED ISSUE #49: Add information text to empty screens
Currently screens are empty and white, if there is no data.
I would be nice to have a short text on screens without data. An empty screen is always bad user experience.

|example from GitHub app|example Google Play|example Expo|
|---|---|---|
|Screenshot_20200513-104755|Screenshot_20220119-181156|Screenshot_20220119-204559|

### PR DESCRIPTION
**Pull request by ardasnturk**
"No content found" text added by default on a page without content.

The EmptyPage.js file has been created in the Components folder. Updated IndexScreen.js and HtmlScreen.js pages. 

home screen events widget | no events for the current day | home screen about item at the bottom | no content
-- | -- | -- | --
 HomeScreen |  Simulator Screen Shot - iPhone 13 - 2022-01-21 at 13 33 27 | HomeScreen  |  Simulator Screen Shot - iPhone 13 - 2022-01-21 at 13 33 37 
HomeScreen | IndexScreen | HomeScreen | HtmlScreen

Fixes #49 

SVA-493

### CODE REVIEW

**Comment by donni106** [comment]
Thanks @ardasnturk for this pull request! 🚀 

At first sight on the screenshots you found the right places, well done. On the other hand, the rendered text is too big and bold and does not fit to our other parts of the app. There is also some top padding missing on the HtmlScreen. We have some components already, that take care of paddings and text sizes correctly according to the whole design of the app. Can you find and use them? I will go through the code soon and point out opportunities at the certain code parts.
Our app supports landscape mode. Please check as well, if the screens look good in shape for that cases.

**Comment by ardasnturk** [comment]
It was a little difficult for me to find them, as I had not been given any orientation about the code before. I have done my best so far, but if you explain the code beforehand, it will be more explanatory for me. When there was no data in the html page, a blank page appeared before me. That's why I edited the code accordingly. Once again, I will look at the codes and try to arrange them according to the design components.

File: src/components/EmptyPage.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    orientation === 'landscape' || dimensions.width > consts.DIMENSIONS.FULL_SCREEN_MAX_WIDTH;
+
+  return (
+    <Text
+      h4
+      h4Style={[styles.titleStyle, needLandscapeStyle && styles.titleStyleLandscape]}
+      containerStyle={[styles.containerStyle, needLandscapeStyle && styles.containerStyleLandscape]}
+    >
+      {title}
+    </Text>
```
Comment (donni106) [inline src/components/EmptyPage.js:21]: there are some `Text` components which are the base for every rendered text in the app. good point, that you have found `react-native-elements` already. `RegularText` is something we use often with some different props available. can you try to use some existing text component here?

File: src/components/EventList.js [hunk 2]
```diff
@@ -58,6 +60,9 @@ export const EventList = ({
           return <LoadingSpinner loading={!listEndReached} />;
         }
 
+        if (!data?.length) {
+          return <EmptyPage title="Im Moment konnte kein anzuzeigender Inhalt gefunden werden." />;
+        }
```
Comment (donni106) [inline src/components/EventList.js:65]: interesting, works, right? without checking deeper i haven't think of puttin it in the footer. have you tried `ListEmptyComponent` of the react native section list? https://reactnative.dev/docs/sectionlist#listemptycomponent

maybe this fits more in our use case?

**Review by donni106 — COMMENTED** [review]
> It was a little difficult for me to find them, as I had not been given any orientation about the code before. I have done my best so far, but if you explain the code beforehand, it will be more explanatory for me. When there was no data in the html page, a blank page appeared before me. That's why I edited the code accordingly. Once again, I will look at the codes and try to arrange them according to the design components.

i understand and know what you mean. you did a good job with finding solutions for the given problem without knowing anything about the big code base.
the pull request is intended now to discuss the best solution we can achieve together.

**Comment by ardasnturk** [comment]
I got a build error because of eslint rules. That's why I had to commit a few times. 

Problems solved:

- Renamed EmptyPage.js to `EmptyMessage.js`.
- Instead of the React Native Elements Text Component, the `RegularText` component written inside the application is used.
- `ListEmptyComponent` has been added to the event listing screen.

File: src/components/EmptyPage.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
+import PropTypes from 'prop-types';
+import React, { useContext } from 'react';
+import { StyleSheet } from 'react-native';
+import { Text } from 'react-native-elements';
+
+import { colors, consts, normalize } from '../config';
+import { OrientationContext } from '../OrientationProvider';
+
+export const EmptyPage = ({ title }) => {
```
[the remark this answers is not in the document]
Comment (donni106) [reply src/components/EmptyPage.js]: It will be `EmptyMessage` now, which works perfectly, as it really just contains the message for empty stuff.

**Comment by donni106** [comment]
> I got a build error because of eslint rules. That's why I had to commit a few times.
> 
> Problems solved:
> 
> * Renamed EmptyPage.js to `EmptyMessage.js`.
> * Instead of the React Native Elements Text Component, the `RegularText` component written inside the application is used.
> * `ListEmptyComponent` has been added to the event listing screen.

@ardasnturk thanks! sounds very good. can you please share screenshots again (portrait and landscape)?

File: src/components/EventList.js [hunk 4]
```diff
@@ -35,14 +36,19 @@ export const EventList = ({
     }
   };
 
-  const renderItem = useRenderItem(QUERY_TYPES.EVENT_RECORDS, navigation, { noSubtitle });
+  const renderItem = useRenderItem(QUERY_TYPES.EVENT_RECORDS, navigation, {
+    noSubtitle
+  });
```
Comment (donni106) [inline src/components/EventList.js:41]: this was linted by you because of some eslint problems occurred within your development environment?

File: src/components/EventList.js [hunk 5]
```diff
@@ -35,14 +36,19 @@ export const EventList = ({
...
   const sectionedData = data.reduce((previous, current) => {
     if (previous[previous?.length - 1]?.title === current.params.details.listDate) {
       previous[previous.length - 1].data.push(current);
       return previous;
     } else {
-      previous.push({ title: current.params.details.listDate, data: [current] });
+      previous.push({
+        title: current.params.details.listDate,
+        data: [current]
+      });
```
Comment (donni106) [inline src/components/EventList.js:51]: this was linted by you as well because of some eslint problems occurred within your development environment?

File: src/screens/HomeScreen.js [hunk 6]
```diff
@@ -145,7 +145,6 @@ export const HomeScreen = ({ navigation, route }) => {
           refreshTimeKey="publicJsonFile-homeCarousel"
         />
         <Widgets widgetConfigs={widgetConfigs} />
-
```
Comment (donni106) [inline src/screens/HomeScreen.js:148]: why was this line removed here? it was helpful to separate the several sections.

File: src/screens/HtmlScreen.js [hunk 7]
```diff
@@ -22,7 +29,8 @@ export const HtmlScreen = ({ navigation, route }) => {
   const [refreshing, setRefreshing] = useState(false);
   const trackScreenViewAsync = useTrackScreenViewAsync();
 
-  if (!query || !queryVariables?.name) return null;
+  if (!query || !queryVariables?.name)
+    return <EmptyMessage title="Im Moment konnte kein anzuzeigender Inhalt gefunden werden." />;
```
Comment (donni106) [inline src/screens/HtmlScreen.js:33]: we try to avoid hardcoded texts and use references to texts from the file `src/config/texts`.
i could think of something like `texts.empty.content` here.
the text itself could be changes to: "Schade, es wurde kein passender Inhalt gefunden, bitte zu einem späteren Zeitpunkt erneut versuchen."

File: src/screens/IndexScreen.js [hunk 8]
```diff
@@ -248,6 +250,11 @@ export const IndexScreen = ({ navigation, route }) => {
                     <DropdownHeader {...{ query, queryVariables, data, updateListData }} />
                   ) : null
                 }
+                ListEmptyComponent={
+                  listItems ? (
+                    <EmptyMessage title="Im Moment konnte kein anzuzeigender Inhalt gefunden werden." />
```
Comment (donni106) [inline src/screens/IndexScreen.js:255]: as mentioned already, we try to avoid hardcoded texts and use references to texts from the file `src/config/texts`.
i could think of something like `texts.empty.list` here.
the text itself could be changes to: "Schade, es wurden keine passenden Einträge gefunden, bitte zu einem späteren Zeitpunkt erneut versuchen."

File: src/components/EmptyMessage.js [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+  );
+};
+
+EmptyMessage.propTypes = {
+  title: PropTypes.string.isRequired
+};
+
+EmptyMessage.defaultProps = {
+  title: ''
+};
```
[the remark this answers is not in the document]
Comment (ardasnturk) [reply src/components/EmptyMessage.js:20]: You're right, it should be required.

File: src/components/EventList.js (hunk 4 above)
Comment (ardasnturk) [reply src/components/EventList.js:41]: I only got ESlint errors over punctuation marks.

File: src/components/EventList.js (hunk 5 above)
Comment (ardasnturk) [reply src/components/EventList.js:51]: I only got ESlint errors over punctuation marks.

File: src/screens/HomeScreen.js (hunk 6 above)
Comment (ardasnturk) [reply src/screens/HomeScreen.js:148]: I guess it's a Syntax issue. I will be more careful about this from now on.

File: src/screens/HtmlScreen.js (hunk 7 above)
Comment (ardasnturk) [reply src/screens/HtmlScreen.js:33]: I will update the code in this way.

File: src/screens/IndexScreen.js (hunk 8 above)
Comment (ardasnturk) [reply src/screens/IndexScreen.js:255]: I will update the code in this way.

**Comment by ardasnturk** [comment]
Update details: 

- Added `text.empty.list` object to `texts.js`,
- Syntax bug fixed,
- Removed defaultProps in `EmptyMessage.js`,
- Message text has been changed.

EventScreen Portrait | EventScreen Landscape | HtmlScreen Portrait | HtmlScreen Landscape
-- | -- | -- | --
| Simulator Screen Shot - iPhone 13 - 2022-01-24 at 15 46 59 | Simulator Screen Shot - iPhone 13 - 2022-01-24 at 15 47 03 | Simulator Screen Shot - iPhone 13 - 2022-01-24 at 15 47 17 | Simulator Screen Shot - iPhone 13 - 2022-01-24 at 15 47 20
EventScreen | EventScreen | HtmlScreen | HtmlScreen

File: src/screens/HtmlScreen.js (hunk 7 above)
Comment (donni106) [reply src/screens/HtmlScreen.js]: you have missed this one, right? it should be a different text here as for the lists.

Comment (ardasnturk) [reply src/screens/HtmlScreen.js]: Yes true. Sorry, I thought it was the same text. I'm updating right now.

**Comment by ardasnturk** [comment]
Update Details:

- Added `texts.empty.content` to `texts.js` file,
- Changed `HtmlScreen.js` empty content message.

File: src/components/EventList.js (hunk 5 above)
Comment (donni106) [reply src/components/EventList.js:51]: maybe some linting issue, where the max line length is setup to 80 at your machine? we use 100.
but no problem, i will check this before merge and correct if necessary. same in EventList.js.

File: src/screens/IndexScreen.js [hunk 10]
```diff
@@ -248,6 +250,7 @@ export const IndexScreen = ({ navigation, route }) => {
                     <DropdownHeader {...{ query, queryVariables, data, updateListData }} />
                   ) : null
                 }
+                ListEmptyComponent={listItems ? <EmptyMessage title={texts.empty.list} /> : null}
```
Comment (donni106) [inline src/screens/IndexScreen.js:253]: istn't `ListEmptyComponent` implicitly checking the data for presence and therefore we could get rid of `listItems ?...` and pass our `EmptyMessage` directly?

File: src/screens/HtmlScreen.js [hunk 11]
```diff
@@ -22,7 +29,7 @@ export const HtmlScreen = ({ navigation, route }) => {
   const [refreshing, setRefreshing] = useState(false);
   const trackScreenViewAsync = useTrackScreenViewAsync();
 
-  if (!query || !queryVariables?.name) return null;
+  if (!query || !queryVariables?.name) return <EmptyMessage title={texts.empty.content} />;
```
Comment (donni106) [inline src/screens/HtmlScreen.js:32]: we are missing screen wrapping components here. compared to the the "main" return of this component, there are following wrappers we should add here:

- `<SafeAreaViewFlex>`
- `<WrapperWithOrientation>`
- `<Wrapper>`

this way, the empty message should be placed more beautiful on the screen.
can you give it a try? otherwise i will dig deeper before merging the branch once the next release will be drafted.

File: src/screens/IndexScreen.js (hunk 10 above)
Comment (ardasnturk) [reply src/screens/IndexScreen.js:253]: At the moment, I tried it as you said. I did not encounter any error. This is how I am sending an update.

File: src/screens/HtmlScreen.js (hunk 11 above)
Comment (ardasnturk) [reply src/screens/HtmlScreen.js:32]: I didn't know I had to write it this way. I'm updating the code right now.

File: src/screens/IndexScreen.js (hunk 10 above)
Comment (ardasnturk) [reply src/screens/IndexScreen.js:253]: Should I use it with `<SafeAreaViewFlex>` ?

Comment (donni106) [reply src/screens/IndexScreen.js:253]: as the list is rendered already inside the wrappers, i think we do not need anything here.

**Comment by donni106** [comment]
> Update Details:
> 
> * Removed `ListItems` control in `IndexScreen.js`,
> * `HtmlScreen.js` blank text wrapped with:
> * * `<SafeAreaViewFlex>`
> * * `<WrapperWithOrientation>`
> * * `<Wrapper>`

sounds good. and how does it look like now? are you satisfied with the new outcome?

**Comment by ardasnturk** [comment]
> > Update Details:
> 
> > 
> 
> > * Removed `ListItems` control in `IndexScreen.js`,
> 
> > * `HtmlScreen.js` blank text wrapped with:
> 
> > * * `<SafeAreaViewFlex>`
> 
> > * * `<WrapperWithOrientation>`
> 
> > * * `<Wrapper>`
> 
> 
> 
> sounds good. and how does it look like now? are you satisfied with the new outcome?

It definitely looks better now.  I'm slowly starting to understand the code.  Thanks.

**Comment by donni106** [comment]
> It definitely looks better now. I'm slowly starting to understand the code. Thanks.

happy to hear this. would you mind sharing screenshots after visual changes, so that i can see whats happening without running the code?

**Comment by donni106** [comment]
@ardasnturk as a further exercise, regarding your screenshot from earlier today, i feel like users do not need filter possibilities for empty lists, because there is no data to filter for.

150785625-e8c4f95c-700b-4c2f-9ac6-95aa1d255f86

list headers could have an additional check on list items and follow the existing logic just in case that there are data.
what do you think? do you know what i mean?

**Comment by ardasnturk** [comment]
> as a further exercise, regarding your screenshot from earlier today, i feel like users do not need filter possibilities for empty lists, because there is no data to filter for.
> 
> 150785625-e8c4f95c-700b-4c2f-9ac6-95aa1d255f86
> 
> list headers could have an additional check on list items and follow the existing logic just in case that there are data. what do you think? do you know what i mean?

My idea is to keep the filter here. 
If there is no data after filtering, I think that the user will not be able to filter again from here. That's why I think it should stay. I hope I could explain.

**Comment by ardasnturk** [comment]
> > It definitely looks better now. I'm slowly starting to understand the code. Thanks.
> 
> happy to hear this. would you mind sharing screenshots after visual changes, so that i can see whats happening without running the code?

EventScreen Portrait | EventScreen Landscape | HtmlScreen Portrait | HtmlScreen Landscape
-- | -- | -- | --
| Simulator Screen Shot - iPhone 13 - 2022-01-24 at 23 33 07 | Simulator Screen Shot - iPhone 13 - 2022-01-24 at 23 33 11  | Simulator Screen Shot - iPhone 13 - 2022-01-24 at 23 32 53 | Simulator Screen Shot - iPhone 13 - 2022-01-24 at 23 32 57
EventScreen | EventScreen | HtmlScreen | HtmlScreen

**Comment by donni106** [comment]
> > as a further exercise, regarding your screenshot from earlier today, i feel like users do not need filter possibilities for empty lists, because there is no data to filter for.
> > 150785625-e8c4f95c-700b-4c2f-9ac6-95aa1d255f86
> > list headers could have an additional check on list items and follow the existing logic just in case that there are data. what do you think? do you know what i mean?
> 
> My idea is to keep the filter here. If there is no data after filtering, I think that the user will not be able to filter again from here. That's why I think it should stay. I hope I could explain.

give it a try in the simulator or on your device. the filter is always there, no matter the filtered data. so if a user accesses the screen, where no data is provided the filter will not bring data anyways. so the user can filter and filter and it always returns zero data. that's why my thoughts were to remove the filter entirely, if there is no data at all. you see what i mean?

File: src/components/EmptyMessage.js [hunk 12]
```diff
@@ -0,0 +1,16 @@
+import PropTypes from 'prop-types';
+import React from 'react';
+
+import { RegularText } from '.';
+
+export const EmptyMessage = ({ title }) => {
+  return (
+    <RegularText small center>
```
Comment (donni106) [inline src/components/EmptyMessage.js:8]: could setting the prop `placeholder` as well makes sense here to reduce strength of the text?

Comment (ardasnturk) [reply src/components/EmptyMessage.js:8]: Sure, I can try. Should I leave the size as `small`?

**Comment by ardasnturk** [comment]
> > > as a further exercise, regarding your screenshot from earlier today, i feel like users do not need filter possibilities for empty lists, because there is no data to filter for.
> > > 150785625-e8c4f95c-700b-4c2f-9ac6-95aa1d255f86
> > > list headers could have an additional check on list items and follow the existing logic just in case that there are data. what do you think? do you know what i mean?
> > 
> > 
> > My idea is to keep the filter here. If there is no data after filtering, I think that the user will not be able to filter again from here. That's why I think it should stay. I hope I could explain.
> 
> give it a try in the simulator or on your device. the filter is always there, no matter the filtered data. so if a user accesses the screen, where no data is provided the filter will not bring data anyways. so the user can filter and filter and it always returns zero data. that's why my thoughts were to remove the filter entirely, if there is no data at all. you see what i mean?


I guess you want such a design. I updated the codes in this way. I don't want to post it to github without your consent. I hope I understood correctly.

Empty Event List | Full Event List | 
-- | -- | 
Simulator Screen Shot - iPhone 13 - 2022-01-27 at 02 01 39 | Simulator Screen Shot - iPhone 13 - 2022-01-27 at 02 01 32
IndexScreen | IndexScreen |

File: src/components/EmptyMessage.js (hunk 12 above)
Comment (donni106) [reply src/components/EmptyMessage.js:8]: good question, ähm, yes, `small` makes quite sense to me.

**Comment by donni106** [comment]
> I guess you want such a design. I updated the codes in this way. I don't want to post it to github without your consent. I hope I understood correctly.
> 
> | Empty Event List | Full Event List | 
> | -- | -- | 
> | Simulator Screen Shot - iPhone 13 - 2022-01-27 at 02 01 39 | Simulator Screen Shot - iPhone 13 - 2022-01-27 at 02 01 32
> | IndexScreen | IndexScreen |

looks very good. i think this is the better user experience!

File: src/components/EmptyMessage.js [hunk 13]
```diff
@@ -1,13 +1,19 @@
...
+import { RegularText, Wrapper, WrapperWithOrientation, SafeAreaViewFlex } from '.';
 
 export const EmptyMessage = ({ title }) => {
   return (
-    <RegularText small center>
-      {title}
-    </RegularText>
+    <SafeAreaViewFlex>
+      <WrapperWithOrientation>
+        <Wrapper>
```
Comment (donni106) [inline src/components/EmptyMessage.js:10]: why is this needed now and moved here? does it still fit for our cases?

File: src/screens/IndexScreen.js [hunk 14]
```diff
@@ -253,6 +253,7 @@ export const IndexScreen = ({ navigation, route }) => {
                 ListEmptyComponent={<EmptyMessage title={texts.empty.list} />}
                 navigation={navigation}
                 data={listItems}
+                noEvent={listItems.length > 0 ? false : true}
```
Comment (donni106) [inline src/screens/IndexScreen.js:256]: i think this new prop can be avoided and does not need to be passed down. at the end there is `ListHeaderComponent={noEvent ? null : ListHeaderComponent}` now. so i have in my mind to directly avoid passing something as `ListHeaderComponent` here if there is no data instead of a new prop to compare later, you know?

Comment (ardasnturk) [reply src/screens/IndexScreen.js:256]: Yes I understand that I don't need the extra prop.

File: src/components/EmptyMessage.js (hunk 13 above)
Comment (donni106) [reply src/components/EmptyMessage.js:10]: all right, now i understand the reason. then i also think it is better be placed here, thanks.

Comment (ardasnturk) [reply src/components/EmptyMessage.js:10]: If wrapping is deleted | If wrapping is not deleted |
|--|--
Simulator Screen Shot - iPhone 13 - 2022-01-27 at 13 27 05 | Simulator Screen Shot - iPhone 13 - 2022-01-27 at 13 08 18 |
IndexScreen.js | IndexScreen.js |

**Comment by ardasnturk** [comment]
Update Details:

- Deleted `noEvent` prop,
- `ListHeaderComponent` is associated with data.

Screenshots:

IndexScreen.js Portrait | IndexScree.js Landscape |
|--|--
Simulator Screen Shot - iPhone 13 - 2022-01-27 at 13 08 18 | Simulator Screen Shot - iPhone 13 - 2022-01-27 at 13 08 23 |
IndexScreen.js | IndexScreen.js |

**Comment by donni106** [comment]
we could think of reusing the image from #440 for empty screens.

**Comment by ardasnturk** [comment]
Update Details:

- Added image for empty event page
- Changed text and image to placeholder

Screenshots:

IndexScreen.js Portrait | IndexScree.js Landscape |
|--|--
Simulator Screen Shot - iPhone 13 - 2022-03-03 at 09 15 24 | Simulator Screen Shot - iPhone 13 - 2022-03-03 at 09 15 16|
IndexScreen.js | IndexScreen.js |

SVA-493

File: src/components/EventList.js (hunk 4 above)
Comment (donni106) [reply src/components/EventList.js:41]: reverted this style right now together with you in pair review

File: src/components/EventList.js (hunk 5 above)
Comment (donni106) [reply src/components/EventList.js:51]: we will keep the new one for better readability of the object

**Review by donni106 — APPROVED** [review]
